main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -1156,6 +1156,163 @@ object GenExpression {
+      // stack: [index, acc, tupleType, tupleType, terms]
+
+      // this is equivalent to the following, where tupleElements* is empty
+      // stack: [index, acc, tupleType, tupleType, tupleElements*, terms]
+
+      for ((tpe, idx) <- tupleElmsTypes.zipWithIndex.reverse) {
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms]
+        visitor.visitInsn(DUP)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, terms]
+        visitor.visitIntInsn(BIPUSH, idx)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, terms, index]
+        // get terms[index]
+        visitor.visitInsn(AALOAD)
+
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, terms[index]]
+        // cast it to a LitTerm
+        visitor.visitTypeInsn(CHECKCAST, JvmName.Runtime.Fixpoint.Term.LitTerm.toInternalName)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, litterm]
+        // call getFunction on it
+        visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.Runtime.Fixpoint.Term.LitTerm.toInternalName, "getFunction",
+          "()Ljava/util/function/Function;", false)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, function]
+        // call the function with a new object as argument
+        visitor.visitTypeInsn(NEW, JvmType.Object.name.toInternalName)
+        // stack: [index, acc, tupleType,  tupleType, tupleElements*, terms, function, object]
+        visitor.visitInsn(DUP)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, function, object, object]
+        // first initialize the object
+        visitor.visitMethodInsn(INVOKESPECIAL, JvmType.Object.name.toInternalName, "<init>", AsmOps.getMethodDescriptor(List(), JvmType.Void), false)
+        // stack: [index, acc, tupleType, tupleType, tupleElements*, terms, function, object]
+        visitor.visitMethodInsn(INVOKEINTERFACE, JvmName.Function.toInternalName, "apply",
+          "(Ljava/lang/Object;)Ljava/lang/Object;", true)
+        visitor.visitTypeInsn(CHECKCAST, JvmName.Runtime.ProxyObject.toInternalName)
+        // Cast the proxy object into an unboxed value if necessary
+        tpe match {
+          case MonoType.Int32 =>
+            // call Object ProxyObject.getValue()
+            visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.Runtime.ProxyObject.toInternalName, "getValue", "()Ljava/lang/Object;", false)
+            // cast Object into java.lang.Integer
+            visitor.visitTypeInsn(CHECKCAST, JvmName.Integer.toInternalName)
+            // call int java.lang.Integer.intValue()
+            visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.Integer.toInternalName, "intValue", "()I", false)
+          case _ =>
+            println(s"Unsupported MonoType: $tpe, letting it in a proxy object")
+            // TODO: is there a way to automatically unbox what can be unboxed? and leave the rest untouched
```

Comment:
Not sure. We have some stuff in AsmOps. Like unboxIfPrim, but I guess for unboxing proxy objects we don't, so we should put it in one central place.